### PR TITLE
DAOS-15262 vos: Fix probe issue in vos iterator

### PR DESCRIPTION
**Pull request by jolivier23**
The checksum scrubber uses vos_iterate to scan everything in the tree, whether visible or not whereas other service ULTs take great care to avoid any epoch range conflict with aggregation and discard services.  This means that any yield in the iterator needs to check to see if the parent objects or keys are still present before continuing.

This function adds a check before probing in the current level for this condition and backs out to higher level to continue, if necessary.  This check is not necessary for visible iterators or aggregation iterators because there are external guarantees that parent trees will never be removed mid-iteration.

Required-githooks: true

Change-Id: Ic1541d9047720d34c060fea19e39c5e8b6fb465c

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate owners.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: src/vos/vos_iterator.c [hunk 1]
```diff
@@ -328,6 +328,82 @@ vos_iter_finish(daos_handle_t ih)
 	return rc || prc;
 }
 
+static inline void
```
Comment (jolivier23) [inline src/vos/vos_iterator.c:331]: Not to reviewers.  Nothing here is different, it just moved here from later in the file

File: src/vos/vos_iterator.c [hunk 2]
```diff
@@ -787,6 +847,19 @@ vos_iterate_internal(vos_iter_param_t *param, vos_iter_type_t type,
 	}
 	read_time = dtx_is_valid_handle(dth) ? dth->dth_epoch : 0 /* unused */;
 probe:
+	if (recursive && type != VOS_ITER_OBJ && iter->it_parent != NULL &&
```
Comment (jolivier23) [inline src/vos/vos_iterator.c:850]: This is the only block where I changed code.

**Comment by jolivier23** [comment]
can somebody check these results?  Is it a known issue or something my patch could cause?

**Comment by wangshilong** [comment]
> can somebody check these results? Is it a known issue or something my patch could cause?

There are two Functions test failure, one telementry test failure which i saw in other PRs, so it shall be not related to this PR.
Anther one is cont_get_acl.py timeout(129 seconds larger than 100seconds), could this PR cause some performance regressions?

**Comment by jolivier23** [comment]
Any slowdown should be specific to the checksum scrubber so I doubt it would cause that.

All of our other iterators are either visible or aggregation.

File: src/vos/vos_iterator.c [hunk 3]
```diff
@@ -787,6 +847,19 @@ vos_iterate_internal(vos_iter_param_t *param, vos_iter_type_t type,
...
+	    (param->ip_flags & (VOS_IT_RECX_VISIBLE | VOS_IT_FOR_PURGE | VOS_IT_FOR_DISCARD)) ==
+		0) {
+		/** If scanning the whole tree, we need to revalidate the parent
+		 * chain and possibly jump back to another level to continue */
+		rc = vos_iter_validate_internal(iter->it_parent);
```
Comment (NiuYawei) [inline src/vos/vos_iterator.c:855]: it's better to put above changes in vos_iter_cb() (in the if(vos_iter_sched_check(iter))), and just convert rc to zero if rc > 0.

File: src/vos/vos_iterator.c [hunk 4]
```diff
@@ -787,6 +847,19 @@ vos_iterate_internal(vos_iter_param_t *param, vos_iter_type_t type,
...
+		if (rc < 0)
+			D_GOTO(out, rc);
+		if (rc > 0 && anchors->ia_probe_level && anchors->ia_probe_level != iter->it_type)
+			D_GOTO(finish, rc = 0); /** The entry doesn't exist anymore at this level */
+		anchors->ia_probe_level = 0;
```
Comment (NiuYawei) [inline src/vos/vos_iterator.c:860]: These dup code could be avoided if we do revalidation in vos_iter_cb().

File: src/vos/vos_iterator.c (hunk 3 above)
Comment (jolivier23) [reply src/vos/vos_iterator.c]: Good idea

File: src/vos/vos_iterator.c (hunk 4 above)
Comment (jolivier23) [reply src/vos/vos_iterator.c]: done

**Review by NiuYawei — APPROVED** [review]

**Review by wangshilong — APPROVED** [review]

**Pull request merged**